Step5 로또(수동) PR 요청 드립니다.

90mansik opened the pull request:
수동 로또 PR 드립니다.!!
잘 부탁 드립니다.

Commit:
docs: 추가 요건사항 정리

- 2등을 위한 추가 보너스 볼을 더 추첨
- 당첨 통계에도 포함되야 한다.

Commit:
test: 상점 가격 가져오기 테스트 클래스 구현

Commit:
feat: 2등 상금 추가로 인한 2등 판단 플래그 추가

Commit:
test: 수익률 검증을 위한 테스트 구현

Commit:
refactor: 수익률 계산식에 2등 계산 추가

Commit:
test: 테스트 검증 수정

Commit:
test: 테스트 검증 수정

Commit:
fix: valueOf 버그 수정

Commit:
refactor: 당첨 숫자 체크 반환 값 변경

AS IS : int
TO BE : PrizeMoney

Commit:
style: 코드 컨벤션 일치화

Commit:
refactor: 로또번호 캐싱화

Commit:
feat: 보너스 번호 기능 추가

Commit:
test: 로또번호 맞춘 갯수 검증 구현

- winLotto에서 계산 하던 당첨 갯수 검증 부분을
  Lotto에 위임 하도록 수정

Commit:
refactor: 로또 맞춘 갯수 가져오기 Lotto에 위임 하도록 리팩토링

- winLotto에서 계산 하던 당첨 갯수 검증 부분을
  Lotto에 위임 하도록 수정

Commit:
style: 불필요한 import, 공백 제거, 멤버변수 final화

Commit:
test: 로또번호 맞춘 갯수 검증 수정

Commit:
refactor: findWinner 메서드 추가

- 로또 검증,계산부분은 모두 WinReport에서 하도록 수정

Commit:
refactor: 로또 계산 메서드를 lotto에 위임

Commit:
refactor: 로또 계산 메서드를 lotto에 위임

Commit:
test: 5단계 요구사항 추가(수동 로또)

Commit:
feat: 5단계 요구사항 추가(수동 로또)

Commit:
fix: 수동 로또가 0건 입력 됬을 경우 문구 노출 버그 수정

Commit:
refactor: PizeMoney에 보너스볼이 있는지 메시지를 보내도록 리팩토링

Commit:
docs: 완료항목 표시

Commit:
style: 완료항목 표시, 코드 포맷팅

minzzang commented on README.md:
```diff
@@ -53,6 +56,6 @@
->   -  [ ] 6.3 보너스 번호를 받아 로또번호에 포함되어 있는지 비교하여 boolean 리턴
->   -  [ ] 6.4 비교 카운트가 5이면서 보너스 번호 매치 값에 따라 2등 3분을 구분하는 로직 구현
+>   -  [x] 6.3 보너스 번호를 받아 로또번호에 포함되어 있는지 비교하여 boolean 리턴
+>   -  [x] 6.4 비교 카운트가 5이면서 보너스 번호 매치 값에 따라 2등 3분을 구분하는 로직 구현
```
요구사항 정리 👍

minzzang commented on src/main/java/domain/Lotto.java:
```diff
@@ -32,6 +32,25 @@ private void isValidSizeNotEqualsLottoNumber(List<LottoNumber> lottoNumbers) {
         }
     }
 
+    public PrizeMoney countCollectNumber(Lotto winLottoNumbers, LottoNumber bonusNumber) {
```
피드백 반영 👍

minzzang commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -6,7 +6,7 @@ public class LottoNumber {
     private static final int LOTTO_MAX_NUMBER = 45;
 
-    int number;
+    private final int number;
```
불변 변경 👍

minzzang commented on src/main/java/view/InputView.java:
```diff
@@ -21,4 +24,21 @@ public static int inputBonusNumber() {
+        sc.nextLine();
+        LottoMachine lottoMachine = new ManualLottoMachine();
+        for(int i=0; i<manualLottoCount; i++){
+          lottos.add(lottoMachine.splitPurchaseLottoNumbers(sc.nextLine()));
```
inputManualLottos인데 입력 받는 부분과 수동 로또를 만드는 부분으로 2개의 책임이 같이 있는데 분리해보면 어떨까요~? 
또한 자동 로또는 LottoMain에서 처리하고 있고 수동 로또는 InputView에서 처리하고 있는데 이 부분은 어떻게 생각하시나요?

90mansik replied:
리뷰어님 말씀하신대로 로또를 입력받는 부분과 만드는 부분에 대해서 책임을 분리 하였습니다.!

minzzang commented on src/main/java/domain/ManualLottoMachine.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public Lottos purchaseLotto(int lottoTicketCount, Lottos lottos) {
+        return null;
```
수동 로또는 splitPurchaseLottoNumbers만 사용하고 자동 로또는 purchaseLotto만 사용해서 LottoMachine 인터페이스의 의미가 크게 없는 것 같습니다. 
lottoMachine.purchaseLotto() 이런 식으로 통일된 메소드를 호출하는건 어떻게 생각하시나요? 수동 로또와 자동 로또에서 호출하는 메소드의  매개변수가 달라 하나의 추상 메소드로 만들기 어려운 경우에는 멤버 변수를 사용해보는 것도 좋을 것 같습니다.

90mansik replied:
리뷰어님 말씀하신대로 lottoMachine.purchaseLotto()  통일된 메소드를 호출 하도록 변경 했습니다.
꼼꼼한 피드백 감사합니다😄

minzzang reviewed (changes requested):
안녕하세요 지용님~!
5단계 미션 잘 수행해주셨네요! :100:

조금 더 고민해보면 좋을 부분이 있어 코멘트 남겼는데 확인 부탁드립니다~ :smile:

그럼 마지막까지 화이팅입니다!  :muscle:

Commit:
test: LottoMachine 인터페이스 통일된 함수 사용 테스트

Commit:
refactor: LottoMachine 인터페이스의 통일된 메서드 구현

AS IS
 - AutoLottoMachine 에서는 puchaseLotto만 사용
 - ManualLottoMachine 에서는 splitPurchaseLottoNumbers만 사용
TO BE
 - 공통된 purchaseLotto를 사용하도록 수정

Commit:
refactor: 메서드 책임 분리

- InputView.inputManualLottos에 수동로또 입력 부, 수동 로또 만드는 부분이 같은 메서드에 존재하여 책임 분리

minzzang commented on src/main/java/domain/LottoMachine.java:
```diff
@@ -1,9 +1,5 @@
-
-    public Lotto splitPurchaseLottoNumbers(String inputWinLottNumbers);
-
+    void purchaseLotto();
```
인터페이스 변경 👍

minzzang commented on src/main/java/view/InputView.java:
```diff
@@ -24,21 +20,25 @@ public static int inputBonusNumber() {
     }
 
-    public static Lottos inputManualLottos(int manualLottoCount, Lottos lottos ){
+    public static void inputManualLottosPrint(){
```
입력 분리 👍

minzzang reviewed (approved):
안녕하세요 지용님~!

마지막 미션 피드백 잘 반영해주셨네요!! :+1:

지금까지 길고 긴 로또 미션 하시느라 정말 수고하셨습니다 ~ :smile:
이번 미션을 통해 TDD와 객체 지향 설계에 많은 도움이 되셨으면 좋겠네요! :pray:

저와의 미션은 여기서 끝이지만 다른 미션 진행하면서도 궁금한 점이나 어려운 부분 있으면
편하게 dm 해주세요 :smile:

남은 미션들도 모두 완주하셔서 원하는 결과 얻기를 바라겠습니다! :bow:
